Automatically detect new Flyway scripts in Dev Mode

### Description

When running in dev mode it would be nice if new Flyway migration scripts (along side any existing migration scripts under `src/main/resources/db/migration`) would be detected and then executed when the user presses the _Migrate_ button in the Flyway Dev UI.

Currently adding a new migration script and pressing _Migrate_ has no effect. The application must be explicitly restarted (e.g. pressing `s` in the console).

### Implementation ideas

_No response_

Not sure if related (I'm a novice with Quarkus) but I am trying to run a Quarkus app in a local container and use [remote development mode](https://quarkus.io/guides/maven-tooling#remote-development-mode) and I'm facing the same error:
```
backend     | 2022-09-20 14:56:35,633 ERROR [io.qua.dep.dev.IsolatedDevModeMain] (vert.x-worker-thread-1) Failed to start quarkus: java.lang.RuntimeException: java.lang.NullPointerException: Cannot invoke "io.quarkus.bootstrap.classloading.ClassPathResource.getUrl()" because "res" is null
backend     | 	at io.quarkus.runner.bootstrap.AugmentActionImpl.runAugment(AugmentActionImpl.java:330)
backend     | 	at io.quarkus.runner.bootstrap.AugmentActionImpl.reloadExistingApplication(AugmentActionImpl.java:265)
backend     | 	at io.quarkus.runner.bootstrap.AugmentActionImpl.reloadExistingApplication(AugmentActionImpl.java:60)
backend     | 	at io.quarkus.deployment.dev.IsolatedDevModeMain.restartApp(IsolatedDevModeMain.java:251)
backend     | 	at io.quarkus.deployment.dev.IsolatedDevModeMain.restartCallback(IsolatedDevModeMain.java:234)
backend     | 	at io.quarkus.deployment.dev.RuntimeUpdatesProcessor.doScan(RuntimeUpdatesProcessor.java:536)
backend     | 	at io.quarkus.deployment.dev.RuntimeUpdatesProcessor.doScan(RuntimeUpdatesProcessor.java:436)
backend     | 	at io.quarkus.vertx.http.runtime.devmode.VertxHttpHotReplacementSetup$4.handle(VertxHttpHotReplacementSetup.java:152)
backend     | 	at io.quarkus.vertx.http.runtime.devmode.VertxHttpHotReplacementSetup$4.handle(VertxHttpHotReplacementSetup.java:139)
backend     | 	at io.vertx.core.impl.ContextBase.lambda$null$0(ContextBase.java:137)
backend     | 	at io.vertx.core.impl.ContextInternal.dispatch(ContextInternal.java:264)
backend     | 	at io.vertx.core.impl.ContextBase.lambda$executeBlocking$1(ContextBase.java:135)
backend     | 	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
backend     | 	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
backend     | 	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1452)
backend     | 	at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
backend     | 	at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
backend     | 	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
backend     | 	at java.base/java.lang.Thread.run(Thread.java:833)
backend     | Caused by: java.lang.NullPointerException: Cannot invoke "io.quarkus.bootstrap.classloading.ClassPathResource.getUrl()" because "res" is null
backend     | 	at io.quarkus.bootstrap.classloading.QuarkusClassLoader.getResources(QuarkusClassLoader.java:250)
backend     | 	at io.quarkus.bootstrap.classloading.QuarkusClassLoader.getResources(QuarkusClassLoader.java:265)
backend     | 	at io.quarkus.bootstrap.classloading.QuarkusClassLoader.getResources(QuarkusClassLoader.java:197)
backend     | 	at io.smallrye.common.classloader.ClassPathUtils.consumeAsPaths(ClassPathUtils.java:84)
backend     | 	at io.smallrye.config.AbstractLocationConfigSourceLoader.tryClassPath(AbstractLocationConfigSourceLoader.java:128)
backend     | 	at io.smallrye.config.AbstractLocationConfigSourceLoader.loadConfigSources(AbstractLocationConfigSourceLoader.java:93)
backend     | 	at io.smallrye.config.AbstractLocationConfigSourceLoader.loadConfigSources(AbstractLocationConfigSourceLoader.java:76)
backend     | 	at io.quarkus.runtime.configuration.ApplicationPropertiesConfigSourceLoader$InClassPath.getConfigSources(ApplicationPropertiesConfigSourceLoader.java:30)
backend     | 	at io.quarkus.runtime.configuration.ApplicationPropertiesConfigSourceLoader$InClassPath.getConfigSources(ApplicationPropertiesConfigSourceLoader.java:27)
backend     | 	at io.smallrye.config.SmallRyeConfigBuilder.build(SmallRyeConfigBuilder.java:439)
backend     | 	at io.quarkus.deployment.ExtensionLoader.loadStepsFrom(ExtensionLoader.java:179)
backend     | 	at io.quarkus.deployment.QuarkusAugmentor.run(QuarkusAugmentor.java:105)
backend     | 	at io.quarkus.runner.bootstrap.AugmentActionImpl.runAugment(AugmentActionImpl.java:328)
backend     | 	... 18 more
```
I am not using any database migration tool. Could it be related?

--- a/extensions/flyway/deployment/src/main/java/io/quarkus/flyway/deployment/FlywayProcessor.java
+++ b/extensions/flyway/deployment/src/main/java/io/quarkus/flyway/deployment/FlywayProcessor.java
@@ -125,9 +125,8 @@ class FlywayProcessor {
         for (var dataSourceName : dataSourceNames) {
             FlywayDataSourceBuildTimeConfig flywayDataSourceBuildTimeConfig = flywayBuildTimeConfig
                     .datasources().get(dataSourceName);
-
             Collection<String> migrationLocations = discoverApplicationMigrations(
-                    flywayDataSourceBuildTimeConfig.locations());
+                    flywayDataSourceBuildTimeConfig.locations(), hotDeploymentProducer);
             applicationMigrationsToDs.put(dataSourceName, migrationLocations);
         }
         Set<String> datasourcesWithMigrations = new HashSet<>();
@@ -312,7 +311,8 @@ class FlywayProcessor {
         return result;
     }
 
-    private Collection<String> discoverApplicationMigrations(Collection<String> locations)
+    private Collection<String> discoverApplicationMigrations(Collection<String> locations,
+            BuildProducer<HotDeploymentWatchedFileBuildItem> hotDeploymentProducer)
             throws IOException {
         LinkedHashSet<String> applicationMigrationResources = new LinkedHashSet<>();
         // Locations can be a comma separated list
@@ -324,6 +324,11 @@ class FlywayProcessor {
             }
 
             String finalLocation = location;
+            // We need to restart/reprocess this if new migration files are added
+            hotDeploymentProducer.produce(HotDeploymentWatchedFileBuildItem.builder()
+                    .setRestartNeeded(true)
+                    .setLocationPredicate(l -> l.startsWith(finalLocation))
+                    .build());
             ClassPathUtils.consumeAsPaths(Thread.currentThread().getContextClassLoader(), location, path -> {
                 Set<String> applicationMigrations = null;
                 try {

--- a/extensions/flyway/deployment/src/main/java/io/quarkus/flyway/deployment/FlywayProcessor.java
+++ b/extensions/flyway/deployment/src/main/java/io/quarkus/flyway/deployment/FlywayProcessor.java
@@ -9,12 +9,13 @@ import java.nio.file.Path;
 import java.nio.file.Paths;
 import java.util.AbstractCollection;
 import java.util.Collection;
-import java.util.HashMap;
+import java.util.Collections;
 import java.util.HashSet;
 import java.util.LinkedHashSet;
 import java.util.List;
 import java.util.Map;
 import java.util.Set;
+import java.util.TreeMap;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
 
@@ -121,26 +122,20 @@ class FlywayProcessor {
             FlywayBuildTimeConfig flywayBuildTimeConfig) throws Exception {
 
         Collection<String> dataSourceNames = getDataSourceNames(jdbcDataSourceBuildItems);
-        Map<String, Collection<String>> applicationMigrationsToDs = new HashMap<>();
+        Map<String, MigrationState> migrationState = new TreeMap<>();
         for (var dataSourceName : dataSourceNames) {
             FlywayDataSourceBuildTimeConfig flywayDataSourceBuildTimeConfig = flywayBuildTimeConfig
                     .datasources().get(dataSourceName);
-            Collection<String> migrationLocations = discoverApplicationMigrations(
-                    flywayDataSourceBuildTimeConfig.locations(), hotDeploymentProducer);
-            applicationMigrationsToDs.put(dataSourceName, migrationLocations);
-        }
-        Set<String> datasourcesWithMigrations = new HashSet<>();
-        Set<String> datasourcesWithoutMigrations = new HashSet<>();
-        for (var e : applicationMigrationsToDs.entrySet()) {
-            if (e.getValue().isEmpty()) {
-                datasourcesWithoutMigrations.add(e.getKey());
-            } else {
-                datasourcesWithMigrations.add(e.getKey());
-            }
+            Set<String> resourcesLocations = new LinkedHashSet<>();
+            Set<String> migrations = new LinkedHashSet<>();
+            discoverApplicationMigrations(flywayDataSourceBuildTimeConfig.locations(), hotDeploymentProducer,
+                    resourcesLocations, migrations);
+            migrationState.put(dataSourceName, new MigrationState(resourcesLocations, migrations));
         }
 
-        Collection<String> applicationMigrations = applicationMigrationsToDs.values().stream().collect(HashSet::new,
-                AbstractCollection::addAll, HashSet::addAll);
+        Collection<String> applicationMigrations = migrationState.values().stream()
+                .map(MigrationState::migrations)
+                .collect(HashSet::new, AbstractCollection::addAll, HashSet::addAll);
         for (String applicationMigration : applicationMigrations) {
             Location applicationMigrationLocation = new Location(applicationMigration);
             String applicationMigrationPath = applicationMigrationLocation.getPath();
@@ -166,7 +161,7 @@ class FlywayProcessor {
         recorder.setApplicationCallbackClasses(callbacks);
 
         resourceProducer.produce(new NativeImageResourceBuildItem(applicationMigrations.toArray(new String[0])));
-        return new MigrationStateBuildItem(datasourcesWithMigrations, datasourcesWithoutMigrations);
+        return new MigrationStateBuildItem(migrationState);
     }
 
     @SuppressWarnings("unchecked")
@@ -202,11 +197,11 @@ class FlywayProcessor {
         Collection<String> dataSourceNames = getDataSourceNames(jdbcDataSourceBuildItems);
 
         for (String dataSourceName : dataSourceNames) {
-            boolean hasMigrations = migrationsBuildItem.hasMigrations.contains(dataSourceName);
-            boolean createPossible = false;
-            if (!hasMigrations) {
-                createPossible = sqlGeneratorBuildItems.stream().anyMatch(s -> s.getDatabaseName().equals(dataSourceName));
-            }
+            var state = migrationsBuildItem.state.get(dataSourceName);
+            boolean hasMigrations = !state.migrations.isEmpty();
+            boolean createPossible = !hasMigrations
+                    && sqlGeneratorBuildItems.stream().anyMatch(s -> s.getDatabaseName().equals(dataSourceName));
+            var ressourcesLocations = state.resourcesLocations;
 
             SyntheticBeanBuildItem.ExtendedBeanConfigurator flywayContainerConfigurator = SyntheticBeanBuildItem
                     .configure(FlywayContainer.class)
@@ -218,7 +213,8 @@ class FlywayProcessor {
                             AgroalDataSourceBuildUtil.qualifier(dataSourceName))
                     .startup()
                     .checkActive(recorder.flywayCheckActiveSupplier(dataSourceName))
-                    .createWith(recorder.flywayContainerFunction(dataSourceName, hasMigrations, createPossible));
+                    .createWith(recorder.flywayContainerFunction(dataSourceName, hasMigrations, createPossible,
+                            ressourcesLocations));
 
             AnnotationInstance flywayContainerQualifier;
 
@@ -288,7 +284,10 @@ class FlywayProcessor {
 
         // once we are done running the migrations, we produce a build item indicating that the
         // schema is "ready"
-        schemaReadyBuildItem.produce(new JdbcDataSourceSchemaReadyBuildItem(migrationsBuildItem.hasMigrations));
+        schemaReadyBuildItem.produce(new JdbcDataSourceSchemaReadyBuildItem(migrationsBuildItem.state.entrySet().stream()
+                .filter(e -> !e.getValue().migrations.isEmpty())
+                .map(Map.Entry::getKey)
+                .toList()));
         initializationCompleteBuildItem.produce(new InitTaskCompletedBuildItem("flyway"));
         return new ServiceStartBuildItem("flyway");
     }
@@ -311,10 +310,11 @@ class FlywayProcessor {
         return result;
     }
 
-    private Collection<String> discoverApplicationMigrations(Collection<String> locations,
-            BuildProducer<HotDeploymentWatchedFileBuildItem> hotDeploymentProducer)
+    private void discoverApplicationMigrations(Collection<String> locations,
+            BuildProducer<HotDeploymentWatchedFileBuildItem> hotDeploymentProducer,
+            Set<String> resourcesLocations,
+            Set<String> applicationMigrationResources)
             throws IOException {
-        LinkedHashSet<String> applicationMigrationResources = new LinkedHashSet<>();
         // Locations can be a comma separated list
         for (String location : locations) {
             location = normalizeLocation(location);
@@ -323,6 +323,7 @@ class FlywayProcessor {
                 continue;
             }
 
+            resourcesLocations.add(location);
             String finalLocation = location;
             // We need to restart/reprocess this if new migration files are added
             hotDeploymentProducer.produce(HotDeploymentWatchedFileBuildItem.builder()
@@ -342,7 +343,6 @@ class FlywayProcessor {
                 }
             });
         }
-        return applicationMigrationResources;
     }
 
     private String normalizeLocation(String location) {
@@ -380,12 +380,13 @@ class FlywayProcessor {
 
     public static final class MigrationStateBuildItem extends SimpleBuildItem {
 
-        final Set<String> hasMigrations;
-        final Set<String> missingMigrations;
+        final Map<String, MigrationState> state;
 
-        MigrationStateBuildItem(Set<String> hasMigrations, Set<String> missingMigrations) {
-            this.hasMigrations = hasMigrations;
-            this.missingMigrations = missingMigrations;
+        MigrationStateBuildItem(Map<String, MigrationState> state) {
+            this.state = Collections.unmodifiableMap(state);
         }
+    }
+
+    public record MigrationState(Set<String> resourcesLocations, Set<String> migrations) {
     }
 }

--- a/extensions/flyway/deployment/src/main/java/io/quarkus/flyway/deployment/devui/FlywayDevUIProcessor.java
+++ b/extensions/flyway/deployment/src/main/java/io/quarkus/flyway/deployment/devui/FlywayDevUIProcessor.java
@@ -41,7 +41,7 @@ public class FlywayDevUIProcessor {
 
         String artifactId = curateOutcomeBuildItem.getApplicationModel().getAppArtifact().getArtifactId();
 
-        recorder.setSqlSuppliers(initialSqlSuppliers, updateSqlSuppliers, artifactId);
+        recorder.initializeJsonRpcService(initialSqlSuppliers, updateSqlSuppliers, artifactId);
 
         CardPageBuildItem card = new CardPageBuildItem();
 

--- a/extensions/flyway/deployment/src/main/resources/dev-ui/qwc-flyway-datasources.js
+++ b/extensions/flyway/deployment/src/main/resources/dev-ui/qwc-flyway-datasources.js
@@ -30,7 +30,6 @@ export class QwcFlywayDatasources extends QwcHotReloadElement {
         _ds: {state: true},
         _selectedDs: {state: true},
         _createDialogOpened: {state: true},
-        _cleanDisabled: {state: true}
     }
     
     constructor() { 
@@ -40,8 +39,7 @@ export class QwcFlywayDatasources extends QwcHotReloadElement {
         this._selectedDs = null;
         this._createDialogOpened = false;
         this._updateDialogOpened = false;
-        this._cleanDisabled = true;
-    }    
+    }
     
     connectedCallback() {
         super.connectedCallback();
@@ -51,10 +49,6 @@ export class QwcFlywayDatasources extends QwcHotReloadElement {
     hotReload(){
         this.jsonRpc.getDatasources().then(jsonRpcResponse => {
             this._ds = jsonRpcResponse.result;
-        });
-        
-        this.jsonRpc.isCleanDisabled().then(jsonRpcResponse => {
-            this._cleanDisabled = jsonRpcResponse.result;
         });
     }
 
@@ -89,15 +83,15 @@ export class QwcFlywayDatasources extends QwcHotReloadElement {
 
     _renderMigrationButtons(ds) {
         if(ds.hasMigrations){
-            let colorvar = this._cleanDisabled ? '--lumo-disabled-text-color' : '--lumo-warning-text-color';
+            let colorvar = ds.cleanDisabled ? '--lumo-disabled-text-color' : '--lumo-warning-text-color';
             return html`<div id=${ds.name} style="display: inline-block;">
-                <vaadin-button theme="small" @click=${() => this._clean(ds)} class="button" ?disabled=${this._cleanDisabled}>
+                <vaadin-button theme="small" @click=${() => this._clean(ds)} class="button" ?disabled=${ds.cleanDisabled}>
                     <vaadin-icon style="color: var(${colorvar});" icon="font-awesome-solid:broom"></vaadin-icon> ${msg('Clean', { id: 'quarkus-flyway-clean' })}
                 </vaadin-button></div>
                 <vaadin-button theme="small" @click=${() => this._migrate(ds)} class="button">
                     <vaadin-icon icon="font-awesome-solid:arrow-right-arrow-left"></vaadin-icon> ${msg('Migrate', { id: 'quarkus-flyway-migrate' })}
                 </vaadin-button>
-                ${this._cleanDisabled
+                ${ds.cleanDisabled
                     ? html`<vaadin-tooltip for="${ds.name}" text=${msg('Flyway clean has been disabled via quarkus.flyway.clean-disabled=true', { id: 'quarkus-flyway-clean-disabled-tooltip' })}></vaadin-tooltip>`
                     : null}
                 `;

--- a/extensions/flyway/deployment/src/test/java/io/quarkus/flyway/test/FlywayDevModeAddMigrationTest.java
+++ b/extensions/flyway/deployment/src/test/java/io/quarkus/flyway/test/FlywayDevModeAddMigrationTest.java
@@ -1,0 +1,53 @@
+package io.quarkus.flyway.test;
+
+import static org.hamcrest.Matchers.is;
+
+import java.sql.Connection;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Statement;
+
+import jakarta.inject.Inject;
+import jakarta.ws.rs.GET;
+import jakarta.ws.rs.Path;
+
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.agroal.api.AgroalDataSource;
+import io.quarkus.test.QuarkusDevModeTest;
+import io.restassured.RestAssured;
+
+// Reproducer for https://github.com/quarkusio/quarkus/issues/25256
+public class FlywayDevModeAddMigrationTest {
+
+    @RegisterExtension
+    static final QuarkusDevModeTest config = new QuarkusDevModeTest()
+            .withApplicationRoot((jar) -> jar
+                    .addClasses(RowCountEndpoint.class)
+                    .addAsResource("db/migration/V1.0.0__Quarkus.sql")
+                    .addAsResource("clean-and-migrate-at-start-config.properties", "application.properties"));
+
+    @Test
+    public void testAddingMigrationScriptCausesRestart() {
+        RestAssured.get("/row-count").then().statusCode(200).body(is("0"));
+        config.addResourceFile("db/migration/V1.1.0__Data.sql", "INSERT INTO quarked_flyway VALUES (1001, 'test')");
+        RestAssured.get("/row-count").then().statusCode(200).body(is("1"));
+    }
+
+    @Path("/row-count")
+    public static class RowCountEndpoint {
+
+        @Inject
+        AgroalDataSource dataSource;
+
+        @GET
+        public int rowCount() throws SQLException {
+            try (Connection connection = dataSource.getConnection(); Statement stat = connection.createStatement()) {
+                try (ResultSet countQuery = stat.executeQuery("select count(1) from quarked_flyway")) {
+                    return countQuery.first() ? countQuery.getInt(1) : 0;
+                }
+            }
+        }
+    }
+}

--- a/extensions/flyway/deployment/src/test/java/io/quarkus/flyway/test/FlywayDevModeCreateFromHibernateTest.java
+++ b/extensions/flyway/deployment/src/test/java/io/quarkus/flyway/test/FlywayDevModeCreateFromHibernateTest.java
@@ -7,8 +7,6 @@ import java.util.function.Function;
 import jakarta.annotation.PostConstruct;
 import jakarta.inject.Inject;
 import jakarta.persistence.EntityManager;
-import jakarta.transaction.Transactional;
-import jakarta.transaction.UserTransaction;
 import jakarta.ws.rs.GET;
 import jakarta.ws.rs.Path;
 
@@ -23,9 +21,11 @@ import org.junit.jupiter.api.extension.RegisterExtension;
 import com.fasterxml.jackson.databind.JsonNode;
 
 import io.quarkus.devui.tests.DevUIJsonRPCTest;
+import io.quarkus.narayana.jta.QuarkusTransaction;
 import io.quarkus.runtime.Startup;
 import io.quarkus.test.QuarkusDevModeTest;
 import io.restassured.RestAssured;
+import io.restassured.filter.log.LogDetail;
 
 public class FlywayDevModeCreateFromHibernateTest extends DevUIJsonRPCTest {
 
@@ -66,7 +66,11 @@ public class FlywayDevModeCreateFromHibernateTest extends DevUIJsonRPCTest {
                 "        return this;\n" +
                 "    }"));
         //added a field, should now fail (if hibernate were still in charge this would work)
-        RestAssured.get("fruit").then().statusCode(500);
+        RestAssured.get("fruit")
+                .then()
+                .log().ifValidationFails(LogDetail.BODY)
+                .statusCode(500)
+                .body(CoreMatchers.containsString("Column \"COLOR\" not found"));
         //now update out sql
         config.modifyResourceFile("db/create/V1.0.0__quarkus-flyway-deployment.sql", new Function<String, String>() {
             @Override
@@ -86,27 +90,18 @@ public class FlywayDevModeCreateFromHibernateTest extends DevUIJsonRPCTest {
         @Inject
         EntityManager entityManager;
 
-        @Inject
-        UserTransaction tx;
-
         @GET
         public List<Fruit> list() {
             return entityManager.createQuery("from Fruit", Fruit.class).getResultList();
         }
 
         @PostConstruct
-        @Transactional
-        public void add() throws Exception {
-            tx.begin();
-            try {
+        public void add() {
+            QuarkusTransaction.requiringNew().run(() -> {
                 Fruit f = new Fruit();
                 f.setName("Orange");
                 entityManager.persist(f);
-                tx.commit();
-            } catch (Exception e) {
-                tx.rollback();
-                throw e;
-            }
+            });
         }
 
     }

--- a/extensions/flyway/deployment/src/test/java/io/quarkus/flyway/test/FlywayDevModeCreateFromHibernateTest.java
+++ b/extensions/flyway/deployment/src/test/java/io/quarkus/flyway/test/FlywayDevModeCreateFromHibernateTest.java
@@ -38,7 +38,8 @@ public class FlywayDevModeCreateFromHibernateTest extends DevUIJsonRPCTest {
             .setArchiveProducer(() -> ShrinkWrap.create(JavaArchive.class)
                     .addClasses(FlywayDevModeCreateFromHibernateTest.class, Endpoint.class, Fruit.class)
                     .addAsResource(new StringAsset(
-                            "quarkus.flyway.locations=db/create"), "application.properties"));
+                            // Trailing spaces are there to check we properly trim the config in the DevUI service.
+                            "quarkus.flyway.locations=db/create   "), "application.properties"));
 
     @Test
     public void testGenerateMigrationFromHibernate() throws Exception {

--- a/extensions/flyway/deployment/src/test/java/io/quarkus/flyway/test/FlywayDevModeCreateFromHibernateTest.java
+++ b/extensions/flyway/deployment/src/test/java/io/quarkus/flyway/test/FlywayDevModeCreateFromHibernateTest.java
@@ -78,9 +78,8 @@ public class FlywayDevModeCreateFromHibernateTest extends DevUIJsonRPCTest {
                 return s + "\nalter table FRUIT add column color VARCHAR;";
             }
         });
-        // TODO: This still fails.
-        //        RestAssured.get("fruit").then().statusCode(200)
-        //                .body("[0].name", CoreMatchers.is("Orange"));
+        RestAssured.get("fruit").then().statusCode(200)
+                .body("[0].name", CoreMatchers.is("Orange"));
     }
 
     @Path("/fruit")

--- a/extensions/flyway/deployment/src/test/java/io/quarkus/flyway/test/FlywayExtensionConfigFixture.java
+++ b/extensions/flyway/deployment/src/test/java/io/quarkus/flyway/test/FlywayExtensionConfigFixture.java
@@ -3,6 +3,8 @@ package io.quarkus.flyway.test;
 import static org.junit.jupiter.api.Assertions.assertEquals;
 
 import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
 import java.util.function.Consumer;
 import java.util.stream.Collectors;
 
@@ -14,6 +16,8 @@ import org.flywaydb.core.Flyway;
 import org.flywaydb.core.api.Location;
 import org.flywaydb.core.api.configuration.Configuration;
 import org.flywaydb.core.api.configuration.FluentConfiguration;
+
+import io.quarkus.flyway.runtime.FlywayContainerUtil;
 
 /**
  * This fixture provides access to read the expected and the actual configuration of flyway.
@@ -61,11 +65,11 @@ public class FlywayExtensionConfigFixture {
     }
 
     public int callbacks(String datasourceName) {
-        return getStringValue("quarkus.flyway.%s.callbacks", datasourceName).split(",").length;
+        return getStringValue("callbacks", datasourceName).split(",").length;
     }
 
     public int connectRetries(String datasourceName) {
-        return getIntValue("quarkus.flyway.%s.connect-retries", datasourceName);
+        return getIntValue("connect-retries", datasourceName);
     }
 
     public int connectRetries(Configuration configuration) {
@@ -73,7 +77,7 @@ public class FlywayExtensionConfigFixture {
     }
 
     public String schemaNames(String datasourceName) {
-        return getStringValue("quarkus.flyway.%s.schemas", datasourceName);
+        return getStringValue("schemas", datasourceName);
     }
 
     public String schemaNames(Configuration configuration) {
@@ -81,7 +85,7 @@ public class FlywayExtensionConfigFixture {
     }
 
     public String tableName(String datasourceName) {
-        return getStringValue("quarkus.flyway.%s.table", datasourceName);
+        return getStringValue("table", datasourceName);
     }
 
     public String tableName(Configuration configuration) {
@@ -89,7 +93,7 @@ public class FlywayExtensionConfigFixture {
     }
 
     public String locations(String datasourceName) {
-        return getStringValue("quarkus.flyway.%s.locations", datasourceName);
+        return getStringValue("locations", datasourceName);
     }
 
     public String locations(Configuration configuration) {
@@ -97,7 +101,7 @@ public class FlywayExtensionConfigFixture {
     }
 
     public String sqlMigrationPrefix(String datasourceName) {
-        return getStringValue("quarkus.flyway.%s.sql-migration-prefix", datasourceName);
+        return getStringValue("sql-migration-prefix", datasourceName);
     }
 
     public String sqlMigrationPrefix(Configuration configuration) {
@@ -105,7 +109,7 @@ public class FlywayExtensionConfigFixture {
     }
 
     public String repeatableSqlMigrationPrefix(String datasourceName) {
-        return getStringValue("quarkus.flyway.%s.repeatable-sql-migration-prefix", datasourceName);
+        return getStringValue("repeatable-sql-migration-prefix", datasourceName);
     }
 
     public String repeatableSqlMigrationPrefix(Configuration configuration) {
@@ -113,7 +117,7 @@ public class FlywayExtensionConfigFixture {
     }
 
     public boolean baselineOnMigrate(String datasourceName) {
-        return getBooleanValue("quarkus.flyway.%s.baseline-on-migrate", datasourceName, false);
+        return getBooleanValue("baseline-on-migrate", datasourceName, false);
     }
 
     public boolean baselineOnMigrate(Configuration configuration) {
@@ -121,7 +125,7 @@ public class FlywayExtensionConfigFixture {
     }
 
     public String baselineVersion(String datasourceName) {
-        return getStringValue("quarkus.flyway.%s.baseline-version", datasourceName);
+        return getStringValue("baseline-version", datasourceName);
     }
 
     public String baselineVersion(Configuration configuration) {
@@ -129,7 +133,7 @@ public class FlywayExtensionConfigFixture {
     }
 
     public String baselineDescription(String datasourceName) {
-        return getStringValue("quarkus.flyway.%s.baseline-description", datasourceName);
+        return getStringValue("baseline-description", datasourceName);
     }
 
     public String baselineDescription(Configuration configuration) {
@@ -137,19 +141,19 @@ public class FlywayExtensionConfigFixture {
     }
 
     public boolean migrateAtStart(String datasourceName) {
-        return getBooleanValue("quarkus.flyway.migrate-at-start", datasourceName, false);
+        return getBooleanValue("migrate-at-start", datasourceName, false);
     }
 
     public String username(String datasourceName) {
-        return getStringValue("quarkus.flyway.%s.username", datasourceName);
+        return getStringValue("username", datasourceName);
     }
 
     public String password(String datasourceName) {
-        return getStringValue("quarkus.flyway.%s.password", datasourceName);
+        return getStringValue("password", datasourceName);
     }
 
     public String jdbcUrl(String datasourceName) {
-        return getStringValue("quarkus.flyway.%s.jdbc-url", datasourceName);
+        return getStringValue("jdbc-url", datasourceName);
     }
 
     private String getStringValue(String parameterName, String datasourceName) {
@@ -172,19 +176,21 @@ public class FlywayExtensionConfigFixture {
         return getValue(parameterName, datasourceName, type, defaultValue, this::log);
     }
 
-    private <T> T getValue(String parameterName, String datasourceName, Class<T> type, T defaultValue, Consumer<String> logger) {
-        String propertyName = fillin(parameterName, datasourceName);
-        T propertyValue = config.getOptionalValue(propertyName, type).orElse(defaultValue);
-        logger.accept("Config property " + propertyName + " = " + propertyValue);
+    private <T> T getValue(String parameterName, String datasourceName, Class<T> type, T defaultValue,
+            Consumer<String> logger) {
+        List<String> propertyNames = FlywayContainerUtil.flywayPropertyKeys(
+                datasourceName == null || datasourceName.isEmpty() ? null : datasourceName, parameterName);
+        Optional<T> propertyValueOptional = Optional.empty();
+        for (String propertyName : propertyNames) {
+            propertyValueOptional = config.getOptionalValue(propertyName, type);
+        }
+        T propertyValue = propertyValueOptional.orElse(defaultValue);
+        logger.accept("Config property " + propertyNames.get(0) + " = " + propertyValue);
         return propertyValue;
     }
 
     private void log(String content) {
         //activate for debugging
         // System.out.println(content);
-    }
-
-    private String fillin(String propertyName, String datasourceName) {
-        return String.format(propertyName, datasourceName).replace("..", ".");
     }
 }

--- a/extensions/flyway/deployment/src/test/java/io/quarkus/flyway/test/FlywayExtensionConfigFixture.java
+++ b/extensions/flyway/deployment/src/test/java/io/quarkus/flyway/test/FlywayExtensionConfigFixture.java
@@ -113,7 +113,7 @@ public class FlywayExtensionConfigFixture {
     }
 
     public boolean baselineOnMigrate(String datasourceName) {
-        return getBooleanValue("quarkus.flyway.%s.baseline-on-migrate", datasourceName);
+        return getBooleanValue("quarkus.flyway.%s.baseline-on-migrate", datasourceName, false);
     }
 
     public boolean baselineOnMigrate(Configuration configuration) {
@@ -137,7 +137,7 @@ public class FlywayExtensionConfigFixture {
     }
 
     public boolean migrateAtStart(String datasourceName) {
-        return getBooleanValue("quarkus.flyway.migrate-at-start", datasourceName);
+        return getBooleanValue("quarkus.flyway.migrate-at-start", datasourceName, false);
     }
 
     public String username(String datasourceName) {
@@ -160,17 +160,21 @@ public class FlywayExtensionConfigFixture {
         return getValue(parameterName, datasourceName, Integer.class);
     }
 
-    private boolean getBooleanValue(String parameterName, String datasourceName) {
-        return getValue(parameterName, datasourceName, Boolean.class);
+    private boolean getBooleanValue(String parameterName, String datasourceName, boolean defaultValue) {
+        return getValue(parameterName, datasourceName, Boolean.class, defaultValue);
     }
 
     private <T> T getValue(String parameterName, String datasourceName, Class<T> type) {
-        return getValue(parameterName, datasourceName, type, this::log);
+        return getValue(parameterName, datasourceName, type, null, this::log);
     }
 
-    private <T> T getValue(String parameterName, String datasourceName, Class<T> type, Consumer<String> logger) {
+    private <T> T getValue(String parameterName, String datasourceName, Class<T> type, T defaultValue) {
+        return getValue(parameterName, datasourceName, type, defaultValue, this::log);
+    }
+
+    private <T> T getValue(String parameterName, String datasourceName, Class<T> type, T defaultValue, Consumer<String> logger) {
         String propertyName = fillin(parameterName, datasourceName);
-        T propertyValue = config.getValue(propertyName, type);
+        T propertyValue = config.getOptionalValue(propertyName, type).orElse(defaultValue);
         logger.accept("Config property " + propertyName + " = " + propertyValue);
         return propertyValue;
     }

--- a/extensions/flyway/runtime-dev/src/main/java/io/quarkus/flyway/runtime/dev/ui/FlywayDevUIRecorder.java
+++ b/extensions/flyway/runtime-dev/src/main/java/io/quarkus/flyway/runtime/dev/ui/FlywayDevUIRecorder.java
@@ -4,19 +4,26 @@ import java.util.Map;
 import java.util.function.Supplier;
 
 import io.quarkus.arc.Arc;
+import io.quarkus.flyway.runtime.FlywayRuntimeConfig;
 import io.quarkus.runtime.RuntimeValue;
 import io.quarkus.runtime.annotations.Recorder;
 
 @Recorder
 public class FlywayDevUIRecorder {
 
-    public RuntimeValue<Boolean> setSqlSuppliers(Map<String, Supplier<String>> initialSqlSuppliers,
+    private final RuntimeValue<FlywayRuntimeConfig> runtimeConfig;
+
+    public FlywayDevUIRecorder(RuntimeValue<FlywayRuntimeConfig> runtimeConfig) {
+        this.runtimeConfig = runtimeConfig;
+    }
+
+    public void initializeJsonRpcService(Map<String, Supplier<String>> initialSqlSuppliers,
             Map<String, Supplier<String>> updateSuppliers, String artifactId) {
         FlywayJsonRpcService rpcService = Arc.container().instance(FlywayJsonRpcService.class).get();
         rpcService.setInitialSqlSuppliers(initialSqlSuppliers);
         rpcService.setUpdateSqlSuppliers(updateSuppliers);
         rpcService.setArtifactId(artifactId);
-        return new RuntimeValue<>(true);
+        rpcService.setConfig(runtimeConfig.getValue());
     }
 
 }

--- a/extensions/flyway/runtime-dev/src/main/java/io/quarkus/flyway/runtime/dev/ui/FlywayJsonRpcService.java
+++ b/extensions/flyway/runtime-dev/src/main/java/io/quarkus/flyway/runtime/dev/ui/FlywayJsonRpcService.java
@@ -21,6 +21,7 @@ import org.flywaydb.core.api.output.MigrateResult;
 import io.quarkus.dev.config.CurrentConfig;
 import io.quarkus.dev.console.DevConsoleManager;
 import io.quarkus.flyway.runtime.FlywayContainer;
+import io.quarkus.flyway.runtime.FlywayContainerUtil;
 import io.quarkus.flyway.runtime.FlywayContainersSupplier;
 import io.quarkus.flyway.runtime.FlywayRuntimeConfig;
 
@@ -153,14 +154,15 @@ public class FlywayJsonRpcService {
                     var dsConfig = runtimeConfig.datasources().get(ds);
                     Map<String, String> newConfig = new HashMap<>();
                     if (dsConfig.baselineOnMigrate().isEmpty()) {
-                        newConfig.put("quarkus.flyway.baseline-on-migrate", "true");
+                        newConfig.put(FlywayContainerUtil.flywayPropertyKey(ds, "baseline-on-migrate"), "true");
                     }
                     if (dsConfig.migrateAtStart().isEmpty()) {
-                        newConfig.put("quarkus.flyway.migrate-at-start", "true");
+                        newConfig.put(FlywayContainerUtil.flywayPropertyKey(ds, "migrate-at-start"), "true");
                     }
                     for (var profile : of("test", "dev")) {
                         if (dsConfig.cleanAtStart().isEmpty()) {
-                            newConfig.put("%" + profile + ".quarkus.flyway.clean-at-start", "true");
+                            newConfig.put("%" + profile + "." + FlywayContainerUtil.flywayPropertyKey(ds, "clean-at-start"),
+                                    "true");
                         }
                     }
                     CurrentConfig.EDITOR.accept(newConfig);

--- a/extensions/flyway/runtime-dev/src/main/java/io/quarkus/flyway/runtime/dev/ui/FlywayJsonRpcService.java
+++ b/extensions/flyway/runtime-dev/src/main/java/io/quarkus/flyway/runtime/dev/ui/FlywayJsonRpcService.java
@@ -11,9 +11,9 @@ import java.util.Collection;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
+import java.util.Set;
 import java.util.function.Supplier;
 
-import org.eclipse.microprofile.config.inject.ConfigProperty;
 import org.flywaydb.core.Flyway;
 import org.flywaydb.core.api.output.CleanResult;
 import org.flywaydb.core.api.output.MigrateResult;
@@ -32,12 +32,6 @@ public class FlywayJsonRpcService {
     private String artifactId;
     private Map<String, FlywayDatasource> datasources;
     private FlywayRuntimeConfig runtimeConfig;
-
-    @ConfigProperty(name = "quarkus.flyway.locations")
-    private List<String> locations;
-
-    @ConfigProperty(name = "quarkus.flyway.clean-disabled")
-    private boolean cleanDisabled;
 
     public void setInitialSqlSuppliers(Map<String, Supplier<String>> initialSqlSuppliers) {
         this.initialSqlSuppliers = initialSqlSuppliers;
@@ -60,15 +54,14 @@ public class FlywayJsonRpcService {
             datasources = new HashMap<>();
             Collection<FlywayContainer> flywayContainers = new FlywayContainersSupplier().get();
             for (FlywayContainer fc : flywayContainers) {
+                var dsName = fc.getDataSourceName();
                 datasources.put(fc.getDataSourceName(),
-                        new FlywayDatasource(fc.getDataSourceName(), fc.isHasMigrations(), fc.isCreatePossible()));
+                        new FlywayDatasource(dsName, fc.isHasMigrations(), fc.isCreatePossible(),
+                                runtimeConfig.datasources().get(dsName).cleanDisabled(),
+                                fc.getResourceLocations()));
             }
         }
         return datasources.values();
-    }
-
-    public boolean isCleanDisabled() {
-        return this.cleanDisabled;
     }
 
     public FlywayActionResponse clean(String ds) {
@@ -126,11 +119,14 @@ public class FlywayJsonRpcService {
         String script = found.get();
 
         Flyway flyway = getFlyway(ds);
+
         if (flyway != null) {
             if (script != null) {
                 try {
+                    FlywayDatasource flywayDatasource = datasources.get(ds);
+                    var locations = flywayDatasource.resourcesLocations;
                     if (locations.isEmpty()) {
-                        return new FlywayActionResponse("error", "Datasource has no locations configured");
+                        return new FlywayActionResponse("error", "Datasource has no locations configured in Java resources");
                     }
 
                     List<Path> resourcesDir = DevConsoleManager.getHotReplacementContext().getResourcesDir();
@@ -141,14 +137,13 @@ public class FlywayJsonRpcService {
                     // In the current project only
                     Path path = resourcesDir.get(0);
 
-                    Path migrationDir = path.resolve(locations.get(0));
+                    Path migrationDir = path.resolve(locations.iterator().next());
                     Files.createDirectories(migrationDir);
                     Path file = migrationDir.resolve(
                             "V1.0.0__" + artifactId + ".sql");
 
                     Files.writeString(file, script);
 
-                    FlywayDatasource flywayDatasource = datasources.get(ds);
                     flywayDatasource.hasMigrations = true;
                     flywayDatasource.createPossible = false;
                     var dsConfig = runtimeConfig.datasources().get(ds);
@@ -188,6 +183,8 @@ public class FlywayJsonRpcService {
     }
 
     public FlywayActionResponse update(String ds) {
+        this.getDatasources(); // Make sure we populated the datasources
+
         try {
             Supplier<String> found = updateSqlSuppliers.get(ds);
             if (found == null) {
@@ -201,8 +198,10 @@ public class FlywayJsonRpcService {
             if (flyway == null) {
                 return errorNoDatasource(ds);
             }
+            FlywayDatasource flywayDatasource = datasources.get(ds);
+            var locations = flywayDatasource.resourcesLocations;
             if (locations.isEmpty()) {
-                return new FlywayActionResponse("error", "Datasource has no locations configured");
+                return new FlywayActionResponse("error", "Datasource has no locations configured in Java resources");
             }
 
             List<Path> resourcesDir = DevConsoleManager.getHotReplacementContext().getResourcesDir();
@@ -212,7 +211,7 @@ public class FlywayJsonRpcService {
             // In the current project only
             Path path = resourcesDir.get(0);
 
-            Path migrationDir = path.resolve(locations.get(0));
+            Path migrationDir = path.resolve(locations.iterator().next());
             Files.createDirectories(migrationDir);
             DateTimeFormatter format = DateTimeFormatter.ofPattern("yyyy.MM.dd.HHmmss");
             String timestamp = LocalDateTime.now().format(format);
@@ -254,18 +253,20 @@ public class FlywayJsonRpcService {
         return null;
     }
 
-    public static class FlywayDatasource {
-        public String name;
+    public static final class FlywayDatasource {
+        public final String name;
         public boolean hasMigrations;
         public boolean createPossible;
+        public final boolean cleanDisabled;
+        public final Set<String> resourcesLocations;
 
-        public FlywayDatasource() {
-        }
-
-        public FlywayDatasource(String name, boolean hasMigrations, boolean createPossible) {
+        public FlywayDatasource(String name, boolean hasMigrations, boolean createPossible, boolean cleanDisabled,
+                Set<String> resourcesLocations) {
             this.name = name;
             this.hasMigrations = hasMigrations;
             this.createPossible = createPossible;
+            this.cleanDisabled = cleanDisabled;
+            this.resourcesLocations = resourcesLocations;
         }
     }
 

--- a/extensions/flyway/runtime-dev/src/main/java/io/quarkus/flyway/runtime/dev/ui/FlywayJsonRpcService.java
+++ b/extensions/flyway/runtime-dev/src/main/java/io/quarkus/flyway/runtime/dev/ui/FlywayJsonRpcService.java
@@ -22,7 +22,7 @@ import io.quarkus.dev.config.CurrentConfig;
 import io.quarkus.dev.console.DevConsoleManager;
 import io.quarkus.flyway.runtime.FlywayContainer;
 import io.quarkus.flyway.runtime.FlywayContainersSupplier;
-import io.quarkus.runtime.configuration.ConfigUtils;
+import io.quarkus.flyway.runtime.FlywayRuntimeConfig;
 
 public class FlywayJsonRpcService {
 
@@ -30,6 +30,7 @@ public class FlywayJsonRpcService {
     private Map<String, Supplier<String>> updateSqlSuppliers;
     private String artifactId;
     private Map<String, FlywayDatasource> datasources;
+    private FlywayRuntimeConfig runtimeConfig;
 
     @ConfigProperty(name = "quarkus.flyway.locations")
     private List<String> locations;
@@ -43,6 +44,10 @@ public class FlywayJsonRpcService {
 
     public void setUpdateSqlSuppliers(Map<String, Supplier<String>> updateSqlSuppliers) {
         this.updateSqlSuppliers = updateSqlSuppliers;
+    }
+
+    public void setConfig(FlywayRuntimeConfig runtimeConfig) {
+        this.runtimeConfig = runtimeConfig;
     }
 
     public void setArtifactId(String artifactId) {
@@ -122,7 +127,6 @@ public class FlywayJsonRpcService {
         Flyway flyway = getFlyway(ds);
         if (flyway != null) {
             if (script != null) {
-                Map<String, String> params = Map.of("ds", ds, "script", script, "artifactId", artifactId);
                 try {
                     if (locations.isEmpty()) {
                         return new FlywayActionResponse("error", "Datasource has no locations configured");
@@ -146,25 +150,30 @@ public class FlywayJsonRpcService {
                     FlywayDatasource flywayDatasource = datasources.get(ds);
                     flywayDatasource.hasMigrations = true;
                     flywayDatasource.createPossible = false;
+                    var dsConfig = runtimeConfig.datasources().get(ds);
                     Map<String, String> newConfig = new HashMap<>();
-                    boolean isBaselineOnMigrateConfigured = ConfigUtils
-                            .isPropertyPresent("quarkus.flyway.baseline-on-migrate");
-                    boolean isMigrateAtStartConfigured = ConfigUtils.isPropertyPresent("quarkus.flyway.migrate-at-start");
-                    boolean isCleanAtStartConfigured = ConfigUtils.isPropertyPresent("quarkus.flyway.clean-at-start");
-                    if (!isBaselineOnMigrateConfigured) {
+                    if (dsConfig.baselineOnMigrate().isEmpty()) {
                         newConfig.put("quarkus.flyway.baseline-on-migrate", "true");
                     }
-                    if (!isMigrateAtStartConfigured) {
+                    if (dsConfig.migrateAtStart().isEmpty()) {
                         newConfig.put("quarkus.flyway.migrate-at-start", "true");
                     }
                     for (var profile : of("test", "dev")) {
-                        if (!isCleanAtStartConfigured) {
+                        if (dsConfig.cleanAtStart().isEmpty()) {
                             newConfig.put("%" + profile + ".quarkus.flyway.clean-at-start", "true");
                         }
                     }
                     CurrentConfig.EDITOR.accept(newConfig);
-                    //force a scan, to make sure everything is up-to-date
-                    DevConsoleManager.getHotReplacementContext().doScan(true);
+                    // TODO ideally we'd force a scan here, but that fails because
+                    //  we end up closing the JsonRpcService that is currently executing...
+                    //  So for now we'll just rely on the next call to any API automatically triggering a restart.
+                    //  Note the line of code below has never actually worked since
+                    //  https://github.com/quarkusio/quarkus/commit/d68896ac48a0e4a6e95b60f3cb3af36e3bd98764#diff-b07030933573531c160a57f757c6949cb3d21dd4132f86656fac2e63413d24d8
+                    //  because:
+                    //  1. Adding a new Flyway file does not result in a restart; see https://github.com/quarkusio/quarkus/issues/25256
+                    //  2. The automatic addition of config was broken until the commit that commented out the line below.
+                    //  For ideas on how to fix this, see https://github.com/quarkusio/quarkus/pull/53884#discussion_r3162481517
+                    // DevConsoleManager.getHotReplacementContext().doScan(true);
                     return new FlywayActionResponse("success",
                             "Initial migration created, Flyway will now manage this datasource");
                 } catch (Throwable t) {

--- a/extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayContainer.java
+++ b/extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayContainer.java
@@ -1,5 +1,7 @@
 package io.quarkus.flyway.runtime;
 
+import java.util.Set;
+
 import org.flywaydb.core.Flyway;
 
 public class FlywayContainer {
@@ -16,11 +18,13 @@ public class FlywayContainer {
     private final String dataSourceName;
     private final boolean hasMigrations;
     private final boolean createPossible;
+    private final Set<String> resourceLocations;
     private final String id;
 
     public FlywayContainer(Flyway flyway, boolean baselineAtStart, boolean cleanAtStart, boolean cleanOnValidationError,
             boolean migrateAtStart, boolean repairAtStart, boolean validateAtStart,
-            String dataSourceName, boolean hasMigrations, boolean createPossible) {
+            String dataSourceName, boolean hasMigrations, boolean createPossible,
+            Set<String> resourceLocations) {
         this.flyway = flyway;
         this.baselineAtStart = baselineAtStart;
         this.cleanAtStart = cleanAtStart;
@@ -31,6 +35,7 @@ public class FlywayContainer {
         this.dataSourceName = dataSourceName;
         this.hasMigrations = hasMigrations;
         this.createPossible = createPossible;
+        this.resourceLocations = resourceLocations;
         this.id = dataSourceName.replace("<", "").replace(">", "");
     }
 
@@ -72,6 +77,10 @@ public class FlywayContainer {
 
     public boolean isCreatePossible() {
         return createPossible;
+    }
+
+    public Set<String> getResourceLocations() {
+        return resourceLocations;
     }
 
     public String getId() {

--- a/extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayContainerProducer.java
+++ b/extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayContainerProducer.java
@@ -51,8 +51,10 @@ public class FlywayContainerProducer {
         final Flyway flyway = new FlywayCreator(matchingRuntimeConfig, matchingBuildTimeConfig, matchingConfigCustomizers(
                 configCustomizerInstances, dataSourceName)).withCallbacks(callbacks)
                 .createFlyway(dataSource);
-        return new FlywayContainer(flyway, matchingRuntimeConfig.baselineAtStart(), matchingRuntimeConfig.cleanAtStart(),
-                matchingRuntimeConfig.cleanOnValidationError(), matchingRuntimeConfig.migrateAtStart(),
+        return new FlywayContainer(flyway, matchingRuntimeConfig.baselineAtStart(),
+                matchingRuntimeConfig.cleanAtStart().orElse(false),
+                matchingRuntimeConfig.cleanOnValidationError(),
+                matchingRuntimeConfig.migrateAtStart().orElse(false),
                 matchingRuntimeConfig.repairAtStart(), matchingRuntimeConfig.validateAtStart(),
                 dataSourceName, hasMigrations,
                 createPossible);

--- a/extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayContainerProducer.java
+++ b/extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayContainerProducer.java
@@ -44,7 +44,7 @@ public class FlywayContainerProducer {
     }
 
     public FlywayContainer createFlyway(DataSource dataSource, String dataSourceName, boolean hasMigrations,
-            boolean createPossible) {
+            boolean createPossible, Set<String> ressourcesLocations) {
         FlywayDataSourceRuntimeConfig matchingRuntimeConfig = flywayRuntimeConfig.datasources().get(dataSourceName);
         FlywayDataSourceBuildTimeConfig matchingBuildTimeConfig = flywayBuildConfig.datasources().get(dataSourceName);
         final Collection<Callback> callbacks = QuarkusPathLocationScanner.callbacksForDataSource(dataSourceName);
@@ -57,7 +57,7 @@ public class FlywayContainerProducer {
                 matchingRuntimeConfig.migrateAtStart().orElse(false),
                 matchingRuntimeConfig.repairAtStart(), matchingRuntimeConfig.validateAtStart(),
                 dataSourceName, hasMigrations,
-                createPossible);
+                createPossible, ressourcesLocations);
     }
 
     private List<FlywayConfigurationCustomizer> matchingConfigCustomizers(

--- a/extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayContainerUtil.java
+++ b/extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayContainerUtil.java
@@ -16,6 +16,25 @@ public final class FlywayContainerUtil {
     private FlywayContainerUtil() {
     }
 
+    public static String flywayPropertyKey(String datasourceName, String radical) {
+        if (DataSourceUtil.isDefault(datasourceName)) {
+            return "quarkus.flyway." + radical;
+        } else {
+            return "quarkus.flyway.\"" + datasourceName + "\"." + radical;
+        }
+    }
+
+    public static List<String> flywayPropertyKeys(String datasourceName, String radical) {
+        if (DataSourceUtil.isDefault(datasourceName)) {
+            return List.of("quarkus.flyway." + radical);
+        } else {
+            // Two possible syntaxes: with or without quotes
+            return List.of(
+                    "quarkus.flyway.\"" + datasourceName + "\"." + radical,
+                    "quarkus.flyway." + datasourceName + "." + radical);
+        }
+    }
+
     public static FlywayContainer getFlywayContainer(String dataSourceName) {
         return Arc.container().instance(FlywayContainer.class,
                 getFlywayContainerQualifier(dataSourceName)).get();

--- a/extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayCreator.java
+++ b/extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayCreator.java
@@ -97,7 +97,7 @@ class FlywayCreator {
             configure.repeatableSqlMigrationPrefix(flywayRuntimeConfig.repeatableSqlMigrationPrefix().get());
         }
         configure.cleanDisabled(flywayRuntimeConfig.cleanDisabled());
-        configure.baselineOnMigrate(flywayRuntimeConfig.baselineOnMigrate());
+        configure.baselineOnMigrate(flywayRuntimeConfig.baselineOnMigrate().orElse(false));
         configure.validateOnMigrate(flywayRuntimeConfig.validateOnMigrate());
         configure.validateMigrationNaming(flywayRuntimeConfig.validateMigrationNaming());
 

--- a/extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayDataSourceRuntimeConfig.java
+++ b/extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayDataSourceRuntimeConfig.java
@@ -105,8 +105,8 @@ public interface FlywayDataSourceRuntimeConfig {
      * true to execute Flyway clean command automatically when the application starts, false otherwise.
      *
      */
-    @WithDefault("false")
-    boolean cleanAtStart();
+    @ConfigDocDefault("`false`")
+    Optional<Boolean> cleanAtStart();
 
     /**
      * true to prevent Flyway clean operations, false otherwise.
@@ -118,8 +118,8 @@ public interface FlywayDataSourceRuntimeConfig {
      * true to execute Flyway automatically when the application starts, false otherwise.
      *
      */
-    @WithDefault("false")
-    boolean migrateAtStart();
+    @ConfigDocDefault("`false`")
+    Optional<Boolean> migrateAtStart();
 
     /**
      * true to execute a Flyway repair command when the application starts, false otherwise.
@@ -148,8 +148,8 @@ public interface FlywayDataSourceRuntimeConfig {
      * Note that this will not automatically call migrate, you must either enable baselineAtStart or programmatically call
      * flyway.migrate().
      */
-    @WithDefault("false")
-    boolean baselineOnMigrate();
+    @ConfigDocDefault("`false`")
+    Optional<Boolean> baselineOnMigrate();
 
     /**
      * true to execute Flyway baseline automatically when the application starts.

--- a/extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayRecorder.java
+++ b/extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/FlywayRecorder.java
@@ -3,6 +3,7 @@ package io.quarkus.flyway.runtime;
 import java.util.Collection;
 import java.util.Locale;
 import java.util.Map;
+import java.util.Set;
 import java.util.function.Function;
 import java.util.function.Supplier;
 
@@ -84,14 +85,16 @@ public class FlywayRecorder {
 
     public Function<SyntheticCreationalContext<FlywayContainer>, FlywayContainer> flywayContainerFunction(String dataSourceName,
             boolean hasMigrations,
-            boolean createPossible) {
+            boolean createPossible,
+            Set<String> ressourcesLocations) {
         return new Function<>() {
             @Override
             public FlywayContainer apply(SyntheticCreationalContext<FlywayContainer> context) {
                 DataSource dataSource = context.getInjectedReference(DataSource.class,
                         AgroalDataSourceUtil.qualifier(dataSourceName));
                 FlywayContainerProducer flywayProducer = context.getInjectedReference(FlywayContainerProducer.class);
-                return flywayProducer.createFlyway(dataSource, dataSourceName, hasMigrations, createPossible);
+                return flywayProducer.createFlyway(dataSource, dataSourceName, hasMigrations, createPossible,
+                        ressourcesLocations);
             }
         };
     }

--- a/extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/UnconfiguredDataSourceFlywayContainer.java
+++ b/extensions/flyway/runtime/src/main/java/io/quarkus/flyway/runtime/UnconfiguredDataSourceFlywayContainer.java
@@ -1,5 +1,7 @@
 package io.quarkus.flyway.runtime;
 
+import java.util.Collections;
+
 import org.flywaydb.core.Flyway;
 
 /**
@@ -12,7 +14,8 @@ public class UnconfiguredDataSourceFlywayContainer extends FlywayContainer {
     private final Throwable cause;
 
     public UnconfiguredDataSourceFlywayContainer(String dataSourceName, String message, Throwable cause) {
-        super(null, false, false, false, false, false, false, dataSourceName, false, false);
+        super(null, false, false, false, false, false, false, dataSourceName, false, false,
+                Collections.emptySet());
         this.message = message;
         this.cause = cause;
     }

